devicemodel/hw/pci/virtio/virtio_gpio.c

```diff
@@ -0,0 +1,942 @@
+/*
+ * Copyright (C) 2019 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <pthread.h>
+#include <ctype.h>
+#include <linux/types.h>
+#include <linux/gpio.h>
+
+#include "acpi.h"
+#include "dm.h"
+#include "pci_core.h"
+#include "mevent.h"
+#include "virtio.h"
+
+/*
+ *  GPIO virtualization architecture
+ *
+ *                +--------------------------+
+ *                |ACRN DM                   |
+ *                |  +--------------------+  |
+ *                |  |                    |  |  virtqueue
+ *                |  |   GPIO mediator    |<-+-----------+
+ *                |  |                    |  |           |
+ *                |  +-+-----+--------+---+  |           |
+ *   User space   +----|-----|--------|------+           |
+ *           +---------+     |        |                  |
+ *           v               v        v                  |
+ *   +----------------+   +-----+   +----------------+   | +---------------+
+ *  -+ /dev/gpiochip0 +---+ ... +---+ /dev/gpiochipN +-----+ UOS           +-
+ *   +                +   +     +   +                +   | +/dev/gpiochip0 +
+ *   +------------+---+   +--+--+   +-------------+--+   | +------+--------+
+ *   Kernel space |          +--------------+     |      |        |
+ *                +--------------------+    |     |      |        |
+ *                                     v    v     v      |        v
+ *   +---------------------+    +---------------------+  |  +-------------+
+ *   |                     |    |                     |  |  |UOS Virtio   |
+ *   |  pinctrl subsystem  |<---+  gpiolib subsystem  |  +->+GPIO Driver  |
+ *   |                     |    |                     |     |             |
+ *   +--------+------------+    +----------+----------+     +-------------+
+ *            |   +------------------------+
+ *            |   |
+ *  ----------|---|----------------------------------------------------------
+ *   Hardware |   |
+ *            v   v
+ *   +------------------+
+ *   |                  |
+ *   | GPIO controllers |
+ *   |                  |
+ *   +------------------+
+ */
+
+static int gpio_debug;
+static FILE *dbg_file;
+#define VIRTIO_GPIO_LOG_INIT do {					\
+	if (gpio_debug) {						\
+		dbg_file = fopen("/tmp/log.gpio", "w+");		\
```

Comment:
The GPIO mediator will only run once since the virtio_gpio_is_active makes GPIO mediator as singleton module.
whatever, I think I need to check dbg_file before fopen, thanks for suggestion, I will fix this.

Reply:
fixed it

Reply:
> The GPIO mediator will only run once since the virtio_gpio_is_active makes GPIO mediator as singleton module.

Thanks for the clarification!